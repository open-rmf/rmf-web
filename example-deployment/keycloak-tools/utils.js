const http = require('http');
const https = require('https');

async function request(url, options, body) {
  options = {
    rejectUnauthorized: false,
    ...options,
  };

  const headers = {};
  if (typeof body === 'object') {
    headers['Content-Type'] = 'application/json';
  }
  options.headers = {
    ...headers,
    ...options.headers,
  };

  const p = new Promise((res, rej) => {
    const req = https.request(url, options, (resp) => {
      resp.body = '';
      resp.on('end', () => {
        if (resp.statusCode < 200 || resp.statusCode >= 300) {
          rej(resp);
        }
        if (resp.headers['content-type'] === 'application/json') {
          resp.body = JSON.parse(resp.body);
        }
        res(resp);
      });
      resp.on('data', (data) => (resp.body += data.toString()));
    });
    if (body) {
      if (typeof body === 'object') {
        req.write(JSON.stringify(body));
      } else {
        req.write(body);
      }
    }
    req.end();
  });
  return p;
}

async function tryRequest(url, options, body) {
  try {
    return await request(url, options, body);
  } catch (e) {
    if (e instanceof http.IncomingMessage) {
      console.error(e.body);
    } else {
      throw e;
    }
  }
}

async function post(url, data) {
  const postData = new URLSearchParams(data);
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
  };
  return request(url, options, postData.toString());
}

const baseUrl = 'https://example.com/auth';
const masterTokenUrl = `${baseUrl}/realms/master/protocol/openid-connect/token`;
const user = 'admin';
const password = 'admin';

async function getToken() {
  const resp = await post(masterTokenUrl, {
    username: user,
    password: password,
    grant_type: 'password',
    client_id: 'admin-cli',
  });
  return resp.body.access_token;
}

module.exports = {
  request,
  tryRequest,
  post,
  getToken,
  baseUrl,
};
