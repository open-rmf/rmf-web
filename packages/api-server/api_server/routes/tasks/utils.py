import rclpy.node
from rmf_task_msgs.msg import TaskSummary as RmfTaskSummary

from ...models.tasks import TaskProgress, TaskSummary


def get_task_progress(
    task_summary: TaskSummary, ros_node: rclpy.node.Node
) -> TaskProgress:
    """
    convert rmf task summary msg to a task
    """
    now = ros_node.get_clock().now().to_msg().sec  # only use sec
    # Generate a progress percentage
    duration = abs(task_summary.end_time.sec - task_summary.start_time.sec)
    # check if is completed
    if task_summary.state == RmfTaskSummary.STATE_COMPLETED:
        progress = "100%"
    # check if it state is queued/cancelled
    elif duration == 0 or (task_summary.state in [0, 4]):
        progress = "0%"
    else:
        percent = int(100 * (now - task_summary.start_time.sec) / float(duration))
        if percent < 0:
            progress = "0%"
        elif percent > 100:
            progress = "Delayed"
        else:
            progress = f"{percent}%"
    return TaskProgress(status=progress)
