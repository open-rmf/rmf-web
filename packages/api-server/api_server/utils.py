def urlpathjoin(*parts) -> str:
    parts = [s.strip('/') for s in parts]
    parts = [s for s in parts if len(s)]
    return '/' + '/'.join(parts)
