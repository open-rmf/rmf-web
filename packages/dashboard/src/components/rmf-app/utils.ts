import * as RmfModels from 'rmf-models';
import { Place } from './place';

export function getPlaces(buildingMap: RmfModels.BuildingMap): Record<string, Place> {
  const places: Record<string, Place> = {};
  for (const level of buildingMap.levels) {
    for (const graphs of level.nav_graphs) {
      for (const vertex of graphs.vertices) {
        if (vertex.name) {
          places[vertex.name] = { level: level.name, vertex };
        }
      }
    }
  }
  return places;
}
