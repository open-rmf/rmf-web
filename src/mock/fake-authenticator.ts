import EventEmitter from 'eventemitter3';
import Authenticator, { AuthenticatorEventType } from '../components/auth/authenticator';
import { User } from '../components/auth/user';

export default class FakeAuthenticator extends EventEmitter<AuthenticatorEventType>
  implements Authenticator {
  readonly user: User = {
    username: 'fakeUser',
  };

  async init(): Promise<void> {}

  login(): Promise<never> {
    throw new Error('Method not implemented.');
  }

  logout(): Promise<never> {
    throw new Error('Method not implemented.');
  }
}
