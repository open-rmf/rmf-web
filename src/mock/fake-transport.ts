import * as RomiCore from '@osrf/romi-js-core-interfaces';
import Debug from 'debug';
import buildingMap from './data/building-map';
import fakeDispenserStates from './data/dispenser-states';
import fakeDoorStates from './data/door-states';
import fakeFleets from './data/fleets';
import fakeLiftStates from './data/lift-states';

const debug = Debug('fake-transport');

export class FakeTransport extends RomiCore.TransportEvents implements RomiCore.Transport {
  name: string = 'fake';

  createPublisher<Message extends unknown>(
    topic: RomiCore.RomiTopic<Message>,
    options?: RomiCore.Options | undefined,
  ): RomiCore.Publisher<Message> {
    return {
      publish: debug,
    };
  }

  subscribe<Message extends unknown>(
    topic: RomiCore.RomiTopic<Message>,
    cb: RomiCore.SubscriptionCb<Message>,
    options?: RomiCore.Options | undefined,
  ): RomiCore.Subscription {
    debug('subscribe:', topic);
    switch (topic) {
      case RomiCore.doorStates: {
        const doorStates = fakeDoorStates();
        setInterval(() => {
          for (const state of Object.values(doorStates)) {
            cb(state as Message);
          }
        }, 1000);
        break;
      }
      case RomiCore.liftStates: {
        const liftStates = fakeLiftStates();
        setInterval(() => {
          for (const state of Object.values(liftStates)) {
            cb(state as Message);
          }
        }, 1000);
        break;
      }
      case RomiCore.fleetStates: {
        const fleets = fakeFleets();
        setInterval(() => {
          for (const fleet of fleets) {
            cb(fleet as Message);
          }
        }, 1000);
        break;
      }
      case RomiCore.dispenserStates: {
        const dispenserStates = fakeDispenserStates();
        setInterval(() => {
          for (const state of Object.values(dispenserStates)) {
            cb(state as Message);
          }
        }, 1000);
        break;
      }
    }

    // TODO remove the skip validation when the type problem of available_floors is solved
    if (topic !== RomiCore.liftStates && topic.topic === 'lift_states') {
      const liftStates = fakeLiftStates();
      setInterval(() => {
        for (const state of Object.values(liftStates)) {
          cb(state as Message);
        }
      }, 1000);
    }

    return {
      unsubscribe: () => {},
    };
  }

  async call<Request extends unknown, Response extends unknown>(
    service: RomiCore.RomiService<Request, Response>,
    req: Request,
  ): Promise<Response> {
    if (service.service === 'get_building_map') {
      return new RomiCore.GetBuildingMap_Response(await buildingMap()) as any;
    }
    throw new Error('not implemented');
  }

  createService<Request extends unknown, Response extends unknown>(
    service: RomiCore.RomiService<Request, Response>,
  ): RomiCore.Service<Request, Response> {
    throw new Error('not implemented');
  }

  destroy(): void {}
}

export default FakeTransport;
