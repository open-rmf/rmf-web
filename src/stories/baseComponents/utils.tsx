import React from 'react';
import * as RomiCore from '@osrf/romi-js-core-interfaces';
import * as L from 'leaflet';

import { colorPalette } from '../../util/css-utils';

export const viewBoxCoords: string = '0 0 25.794363144785166 14.53525484725833';
export const footprint: number = 0.5;
export const mapBound = new L.LatLngBounds([0, 25.794363144785166], [-17.53525484725833, 0]);
export const maxBound = new L.LatLngBounds(
  [12.84386068880558, 338.5063539594412],
  [-77.06316413283348, -56.417725659906864],
);

export interface StyleTyping {
  [key: string]: React.CSSProperties;
}

export interface FormProps {
  errorState: boolean[];
  errorMessage: string[];
  labels: string[];
}

/***
 * Start of Door utils
 * Use door if you just need a door and doors if you need an array of doors
 * If you need just one door state and not the whole object, use doorStates.main_door
 * The above applies to the Dispenser, Lift and Robot util state
 */

const door: RomiCore.Door = {
  door_type: RomiCore.Door.DOOR_TYPE_SINGLE_SLIDING,
  motion_direction: 1,
  motion_range: -1.571,
  name: 'main_door',
  v1_x: 10.8,
  v1_y: -2.3,
  v2_x: 7.7,
  v2_y: -5.5,
};

const doors: RomiCore.Door[] = [
  { ...door },
  { ...door, name: 'second_door' },
  { ...door, name: 'third_door' },
  { ...door, name: 'fourth_door' },
];

const doorStateTemplate: RomiCore.DoorState = {
  current_mode: {
    value: RomiCore.DoorMode.MODE_CLOSED,
  },
  door_name: 'main_door',
  door_time: { sec: 0, nanosec: 0 },
};

const doorStates: { [key: string]: RomiCore.DoorState } = {
  main_door: {
    ...doorStateTemplate,
  },
  second_door: {
    ...doorStateTemplate,
    current_mode: {
      value: RomiCore.DoorMode.MODE_OPEN,
    },
    door_name: 'second_door',
  },
  third_door: {
    ...doorStateTemplate,
    current_mode: {
      value: RomiCore.DoorMode.MODE_MOVING,
    },
    door_name: 'third_door',
  },
  fourth_door: {
    ...doorStateTemplate,
    current_mode: {
      value: 3,
    },
    door_name: 'fourth_door',
  },
};

// end of door utils

/***
 * Start of Dispenser Utils
 */
const dispenserStateTemplate: RomiCore.DispenserState = {
  guid: 'main_dispenser',
  mode: RomiCore.DispenserState.IDLE,
  request_guid_queue: [],
  seconds_remaining: 0,
  time: { sec: 0, nanosec: 0 },
};

const dispenserStates: { [key: string]: RomiCore.DispenserState } = {
  main_dispenser: { ...dispenserStateTemplate },
  second_dispenser: {
    ...dispenserStateTemplate,
    guid: 'second_dispenser',
    mode: RomiCore.DispenserState.BUSY,
  },
  third_dispenser: {
    ...dispenserStateTemplate,
    guid: 'third_dispenser',
    mode: RomiCore.DispenserState.OFFLINE,
  },
  fourth_dispenser: {
    ...dispenserStateTemplate,
    guid: 'fourth_dispenser',
    mode: 3,
  },
};

// end of dispenser utils

/***
 * Start of lift utils
 */

const lift: RomiCore.Lift = {
  depth: 2.5,
  doors: [
    {
      door_type: 1,
      motion_direction: 1,
      motion_range: 0,
      name: 'main_lift_front_door',
      v1_x: 8.8,
      v1_y: -2.3,
      v2_x: 7.7,
      v2_y: -4.5,
    },
  ],
  levels: ['L1', 'L2', 'L3', 'L4'],
  name: 'main_lift',
  ref_x: 7.1,
  ref_y: -2.8,
  ref_yaw: 0.5,
  wall_graph: {
    edges: [],
    name: 'wallgraph',
    params: [],
    vertices: [],
  },
  width: 2.5,
};

const lifts: RomiCore.Lift[] = [
  { ...lift },
  {
    ...lift,
    doors: [{ ...lift.doors[0], name: 'second_lift_front_door' }],
    name: 'second_lift',
  },
];

const liftStateTemplate: RomiCore.LiftState = {
  available_floors: ['L1', 'L2', 'L3', 'L4'],
  available_modes: new Uint8Array(0),
  current_floor: 'L1',
  current_mode: RomiCore.LiftState.MODE_UNKNOWN,
  destination_floor: 'L1',
  door_state: RomiCore.LiftState.DOOR_CLOSED,
  lift_name: 'main_lift',
  lift_time: { sec: 0, nanosec: 0 },
  motion_state: RomiCore.LiftState.MOTION_STOPPED,
  session_id: '',
};

const liftStates: { [key: string]: RomiCore.LiftState } = {
  main_lift: { ...liftStateTemplate },
  second_lift: {
    ...liftStateTemplate,
    lift_name: 'second_lift',
    motion_state: RomiCore.LiftState.MOTION_UP,
  },
};

// End of lift utils

/***
 * Start of Robot utils
 */
const robotState: RomiCore.RobotState = {
  battery_percent: 100,
  location: {
    level_name: 'L1',
    t: { sec: 0, nanosec: 0 },
    x: 8,
    y: -4,
    yaw: 0,
  },
  mode: { mode: RomiCore.RobotMode.MODE_IDLE },
  model: '40_hours',
  name: 'main_robot',
  path: [],
  task_id: 'taskA',
};

const robotStates: RomiCore.FleetState[] = [
  {
    name: 'main_fleet',
    robots: [
      { ...robotState },
      {
        ...robotState,
        name: 'second_robot',
        mode: { mode: RomiCore.RobotMode.MODE_CHARGING },
      },
      {
        ...robotState,
        name: 'third_robot',
        mode: { mode: RomiCore.RobotMode.MODE_DOCKING },
      },
      {
        ...robotState,
        name: 'fourth_robot',
        mode: { mode: RomiCore.RobotMode.MODE_EMERGENCY },
      },
      {
        ...robotState,
        name: 'fifth_robot',
        mode: { mode: RomiCore.RobotMode.MODE_GOING_HOME },
      },
      {
        ...robotState,
        name: 'sixth_robot',
        mode: { mode: RomiCore.RobotMode.MODE_MOVING },
      },
      {
        ...robotState,
        name: 'seventh_robot',
        mode: { mode: RomiCore.RobotMode.MODE_PAUSED },
      },
      {
        ...robotState,
        name: 'eigth_robot',
        mode: { mode: RomiCore.RobotMode.MODE_WAITING },
      },
    ],
  },
];

// end of Robot utils

export const componentDisplayStyle: StyleTyping = {
  display: {
    display: 'grid',
    gridTemplateColumns: '1fr 3fr',
  },
  modeInfoPanel: {
    padding: '2rem',
  },
  modeInfoItem: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: '0.5rem',
  },
  modeInfoLink: {
    marginTop: '0.5rem',
    padding: '0.5rem',
  },
  aTag: {
    textDecoration: 'none',
    color: 'rgb(20, 116, 243)',
  },
};

export const defaultStyles: StyleTyping = {
  root: {
    margin: '0 auto',
    width: '40%',
  },
  heading: {
    padding: '0.5rem',
  },
  aTag: {
    textDecoration: 'none',
    color: 'rgb(20, 116, 243)',
  },
};

// colors used outside material ui
export const colorPaletteUtil: StyleTyping = {
  unknown: {
    backgroundColor: colorPalette.unknown,
  },
};

export {
  door,
  doors,
  doorStates,
  dispenserStates,
  lift,
  lifts,
  liftStates,
  robotState,
  robotStates,
};
