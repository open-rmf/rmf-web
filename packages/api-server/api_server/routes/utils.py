import asyncio

from rx import operators as rxops
from rx.core.observable.observable import Observable
from rx.scheduler.eventloop.asyncioscheduler import AsyncIOScheduler

from api_server.fast_io import WatchRequest


def rx_watcher(req: WatchRequest, target: Observable):
    def handle(data):
        async def emit(data):
            await req.emit(data)

        loop.create_task(emit(data))

    loop = asyncio.get_event_loop()
    sub = target.pipe(rxops.observe_on(AsyncIOScheduler(loop))).subscribe(handle)
    req.on_unsubscribe(sub.dispose)
