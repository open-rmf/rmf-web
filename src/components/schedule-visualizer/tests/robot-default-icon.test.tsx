import { mount } from 'enzyme';
import React from 'react';
import { makeRobot } from '../../../mock/utils';
import ColorManager from '../colors';
import RobotDefaultIcon from '../robot-default-icon';

describe('RobotDefaultIcon', () => {
  const colorManager = new ColorManager();
  // TextEncoder is not available in node
  colorManager.robotColor = jest.fn(async () => 'black');
  colorManager.robotColorFromCache = jest.fn(() => 'black');

  test('renders correctly', async () => {
    const root = mount(
      <svg>
        <RobotDefaultIcon robot={makeRobot()} footprint={1} colorManager={colorManager} />
      </svg>,
    );
    expect(root).toMatchSnapshot();
  });

  test('red shadow appears for conflicting robots', async () => {
    const root = mount(
      <svg>
        <RobotDefaultIcon
          robot={makeRobot()}
          footprint={1}
          colorManager={colorManager}
          inConflict={true}
        />
      </svg>,
    );
    expect(root.find('circle#shadow[fill*="conflict"]')).toHaveLength(1);
  });
});
