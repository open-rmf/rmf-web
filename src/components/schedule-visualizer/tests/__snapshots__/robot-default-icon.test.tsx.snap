// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RobotDefaultIcon renders correctly 1`] = `
<svg>
  <ForwardRef
    colorManager={
      ColorManager {
        "_robotColorCache": Object {},
        "conflictHighlight": "#f44336",
        "robotColor": [MockFunction],
        "robotColorFromCache": [MockFunction] {
          "calls": Array [
            Array [
              "",
              "BaseModel",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "black",
            },
          ],
        },
      }
    }
    footprint={1}
    robot={
      Object {
        "battery_percent": 100,
        "location": Object {
          "level_name": "",
          "t": Object {
            "nanosec": 0,
            "sec": 0,
          },
          "x": 0,
          "y": 0,
          "yaw": 0,
        },
        "mode": Object {
          "mode": 0,
        },
        "model": "BaseModel",
        "name": "",
        "path": Array [],
        "task_id": "",
      }
    }
  >
    <g>
      <defs>
        <Component
          id="RobotDefaultIcon-0-shadow"
        >
          <radialGradient
            id="RobotDefaultIcon-0-shadow"
          >
            <stop
              offset="70%"
              stopColor="#000000ff"
            />
            <stop
              offset="75%"
              stopColor="#00000080"
            />
            <stop
              offset="80%"
              stopColor="#00000060"
            />
            <stop
              offset="85%"
              stopColor="#00000030"
            />
            <stop
              offset="90%"
              stopColor="#00000018"
            />
            <stop
              offset="95%"
              stopColor="#00000008"
            />
            <stop
              offset="100%"
              stopColor="#00000000"
            />
          </radialGradient>
        </Component>
        <Component
          id="RobotDefaultIcon-0-shadow-conflict"
        >
          <radialGradient
            id="RobotDefaultIcon-0-shadow-conflict"
          >
            <stop
              offset="70%"
              stopColor="#f44336ff"
            />
            <stop
              offset="75%"
              stopColor="#f4433680"
            />
            <stop
              offset="80%"
              stopColor="#f4433660"
            />
            <stop
              offset="85%"
              stopColor="#f4433630"
            />
            <stop
              offset="90%"
              stopColor="#f4433618"
            />
            <stop
              offset="95%"
              stopColor="#f4433608"
            />
            <stop
              offset="100%"
              stopColor="#f4433600"
            />
          </radialGradient>
        </Component>
      </defs>
      <circle
        fill="url(#RobotDefaultIcon-0-shadow)"
        id="shadow"
        r={1.3}
      />
      <circle
        fill="black"
        r={1}
      />
      <line
        stroke="#000"
        strokeWidth="0.05"
        x2={1}
      />
    </g>
  </ForwardRef>
</svg>
`;
